csi.proto

```diff
@@ -1912,3 +1930,172 @@ message GetVolumeGroupSnapshotResponse {
+  // This field is OPTIONAL.
+  int32 max_results = 3;
+
+  // Secrets required by plugin to complete the request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 4 [(csi_secret) = true];
+}
+
+// GetMetadataAllocatedResponse messages are returned in a gRPC stream.
+// Cumulatively, they provide information on the allocated data
+// ranges in the snapshot.
+message GetMetadataAllocatedResponse {
+  // This specifies the style used in the BlockMetadata sequence.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  // If block_metadata_type is FIXED_LENGTH, then the size_bytes field
+  // of each message in the block_metadata list MUST be constant.
+  // This field is REQUIRED.
+  BlockMetadataType block_metadata_type = 1;
+
+  // This returns the capacity of the underlying volume in bytes.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  // This field is REQUIRED.
+  int64 volume_capacity_bytes = 2;
+
+  // This is a list of data range tuples.
+  // If the value of max_results in the GetMetadataAllocatedRequest
+  // message is greater than zero, then the number of entries in this
+  // list MUST be less than or equal to that value.
+  // The byte_offset field of each message MUST be strictly increasing.
+  // The computation of the data range in the first BlockMetadata
+  // message in the stream of GetMetadataAllocatedResponse messages
+  // returned MUST consider the starting_offset requested in the
+  // GetMetadataAllocatedRequest message, though it is not required
+  // that the first data range tuple returned actually include the
+  // starting_offset, as the first allocated block range may start
+  // after this offset.
+  // BlockMetadata messages MUST NOT overlap for a given snapshot, i.e.
+  // for any two BlockMetadata messages, A and B, if
+  // A.byte_offset < B.byte_offset then
+  // A.byte_offset + A.size_bytes <= B.byte_offset MUST be true.
+  // This field is OPTIONAL.
+  repeated BlockMetadata block_metadata = 3;
+}
+
+// The GetMetadataDeltaRequest message is used to solicit metadata on
+// the data ranges that have changed between two snapshots.
+message GetMetadataDeltaRequest {
+  // This is the identifier of the snapshot against which changes
+  // are to be computed.
+  // This field is REQUIRED.
+  string base_snapshot_id = 1;
+
+  // This is the identifier of a second snapshot in the same volume,
+  // created after the base snapshot.
+  // This field is REQUIRED.
+  string target_snapshot_id = 2;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed. It is intended
+  // to be used to continue a previously interrupted call.
+  // The plugins may round down this offset to the nearest alignment
+  // boundary based on the BlockMetadataType used.
```

Comment:
// This indicates the zero based starting byte position in the volume
  // snapshot from which the result should be computed. It is intended
  // to be used to continue a previously interrupted call.
  // The CO SHOULD specify this value to the be the byte immediately
  // after the last byte of the last block received, if continuing an
  // interrupted operation, or zero if not.
  // The SP MUST NOT return any blocks for which the block byte_offset
  // plus the block size_bytes is less than starting_offset.
  // THE SP MAY return blocks with byte_offset less than starting_offset if
  // starting_offset is less than byte_offset plus size_bytes for that block.

Reply:
Cool! I'll generalize this to a "data range" (instead of "block").